File: skellig-junit-runner/src/test/java/org/skellig/runner/config/SimpleMessageTestStepFactory.java
package org.skellig.runner.config;

import org.skellig.teststep.processing.converter.TestDataConverter;
import org.skellig.teststep.processing.converter.TestStepValueConverter;
import org.skellig.teststep.processing.model.factory.BaseTestStepFactory;

import java.util.Map;
import java.util.Properties;

public class SimpleMessageTestStepFactory extends BaseTestStepFactory {

    public SimpleMessageTestStepFactory(Properties keywordsProperties, TestStepValueConverter testStepValueConverter,
                                        TestDataConverter testDataConverter) {
        super(keywordsProperties, testStepValueConverter, testDataConverter);
    }

    @Override
    protected CreateTestStepDelegate createTestStep(Map<String, Object> rawTestStep) {
        return (id, name, testData, validationDetails, parameters, variables) ->
                new SimpleMessageTestStep.Builder()
                        .withReceiver(convertValue(rawTestStep.get("receiver"), parameters))
                        .withId(id)
                        .withName(name)
                        .withVariables(variables)
                        .withTestData(testData)
                        .withValidationDetails(validationDetails)
                        .build();
    }

    @Override
    public boolean isConstructableFrom(Map<String, Object> rawTestStep) {
        return rawTestStep.containsKey("receiver");
    }
}

File: skellig-junit-runner/src/test/kotlin/org/skellig/runner/config/SimpleMessageTestStepFactory.kt
package org.skellig.runner.config

import org.skellig.teststep.processing.model.DefaultTestStep
import org.skellig.teststep.processing.model.factory.BaseDefaultTestStepFactory
import org.skellig.teststep.processing.model.factory.TestStepRegistry
import org.skellig.teststep.processing.value.ValueExpressionContextFactory
import org.skellig.teststep.reader.value.expression.AlphanumericValueExpression
import org.skellig.teststep.reader.value.expression.ValueExpression

class SimpleMessageTestStepFactory(
    testStepRegistry: TestStepRegistry,
    valueExpressionContextFactory: ValueExpressionContextFactory,
    defaultTestDataConverter: String? = null
) : BaseDefaultTestStepFactory<SimpleMessageTestStep>(testStepRegistry, valueExpressionContextFactory, defaultTestDataConverter) {

    companion object {
        private val RECEIVER = fromProperty("receiver")
        private val RECEIVE_FROM = fromProperty("receiveFrom")
    }

    override fun createTestStepBuilder(rawTestStep: Map<ValueExpression, ValueExpression?>, parameters: Map<String, Any?>): DefaultTestStep.Builder<SimpleMessageTestStep> {
        return SimpleMessageTestStep.Builder()
            .withReceiver(convertValue(rawTestStep[RECEIVER], parameters))
            .withReceiveFrom(convertValue(rawTestStep[RECEIVE_FROM], parameters))
    }

    override fun isConstructableFrom(rawTestStep: Map<ValueExpression, ValueExpression?>): Boolean {
        return rawTestStep.containsKey(RECEIVER) || rawTestStep.containsKey(RECEIVE_FROM)
    }
}